_posts/2022-09-28-ltr-with-opensearch-and-metarank.markdown

```diff
@@ -0,0 +1,268 @@
+* use item metadata from stored fields in the ranking (item price, for example)
+* put external ranking factors into the scoring function, like having a separate set of weights for mobile and desktop visitors.
+
+But `script_score` approach also has a set of important drawbacks:
+* As the scoring function is invoked on each matching document, **you cannot do any complex computations inside**.
+* Which scoring formula should you have? There are no common ones, and you have to **invent your own scoring function**.
+* Scoring formula typically has a set of constants and weights. **How to choose the best values for parameters**?
+* Feature values used in the ranking formula should be either stored in the index (**requiring a frequent full reindexing**), or served outside the search (**requiring a custom satellite [Feature Store](https://www.featurestore.org/what-is-a-feature-store)**)
+
+## Secondary reranking
+
+There is another approach to ranking - a secondary/multi-level re-ranking. The idea of such an approach is to split retrieval and multiple re-ranking phases into separate independent steps:
+* first-level fast ranker retrieves a full set of results and uses an algorithm like BM25 to do the initial scoring. This approach reduces the number of matching documents to a top-N most relevant documents, focusing on [recall metric](https://en.wikipedia.org/wiki/Precision_and_recall).
+* second-level slow ranker reorders only top-N candidates, focusing on a [precision metric](https://en.wikipedia.org/wiki/Precision_and_recall).
+
+This approach is widespread in the industry. For a good overview, watch a talk [Berlin Buzzwords 2019: Michael Sokolov & Mike McCandless–E-Commerce search at scale on Apache Lucene](https://youtu.be/EkkzSLstSAE?t=2191) on how multi-stage ranking is implemented on amazon.com:
+
+![Multi-phase ranking]({{ site.baseurl }}/assets/media/blog-images/2022-09-28-ltr-with-opensearch-and-metarank/mphase.png){:class="img-centered"}
+
+Multi-stage ranking also has a set of its tradeoffs:
+* **it operates only on top-N candidates**, so if your first-level ranker under-scored a document (it wasn't included in the top-N set), it will be missing from the final ranking.
+* you had to **build such a system in-house**, as there were no open-source solutions available.
+
+But as final re-ranking is happening outside your search application and happens only on a top-N subset of documents, you can implement complex [Learn-to-Rank](https://en.wikipedia.org/wiki/Learning_to_rank) algorithms like [LambdaMART](https://softwaredoug.com/blog/2021/11/28/how-lammbamart-works.html).
+
+## Metarank
+
+[Metarank](https://metarank.ai) is an open-source personalized secondary reranker:
+* Implements [LambdaMART](https://docs.metarank.ai/reference/overview/supported-ranking-models) on top of embedded feature engineering pipeline.
```

Comment:
Does Metarank only support LambdaMART? If not, could the user indicate the algorithm? Does Metarank support custom models, especially for deep learning models?

Reply:
Good catch. We only implement LambdaMART now, but we do have a plan to try something more trendy in the future. But the problem with lmart is that it's often good enough :)

Added a couple of sentences on the future plans.